crates/datafusion_ext/src/runtime/group_pull_up.rs

```diff
@@ -23,6 +24,54 @@ impl RuntimeGroupPullUp {
     pub fn new() -> Self {
         RuntimeGroupPullUp {}
     }
+
+    /// if a union consists of multiple runtimes, we can rewrite it as 2 groupings
+    /// union[remote, local, remote, remote] -> [union[union[remote, remote, remote], local]]
+    fn rewrite_union(
+        self,
+        plan: Arc<dyn ExecutionPlan>,
+        children: Vec<RuntimeGroupExec>,
+        config: &ConfigOptions,
+    ) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+        let handle_children = |children: Vec<RuntimeGroupExec>| -> Result<Arc<dyn ExecutionPlan>> {
+            if children.len() == 1 {
+                self.optimize(children.get(0).unwrap().clone().child, config)
+            } else {
+                Ok(Arc::new(UnionExec::new(
+                    children
+                        .into_iter()
+                        .map(|x| self.optimize(x.child, config))
+                        .collect::<Result<_>>()?,
+                )))
+            }
+        };
+
+        let it = children.into_iter();
+        let mut remote = Vec::new();
+        let mut local = Vec::new();
+
+        for child in it {
+            match child.preference {
+                RuntimePreference::Remote => remote.push(child),
+                RuntimePreference::Local => local.push(child),
+                RuntimePreference::Unspecified => todo!("unspecified runtime preference"),
+            }
+        }
+
+        // if the plan only contains one runtime, we can just return it as is.
+        if (local.len() | remote.len()) == 0 {
```

Comment:
I'd rewrite this to `local.is_empty() || remote.is_empty()` for clarity.

Reply:
so the logic for this was actually incorrect, the previous tests actually caught it! should be all fixed now!